### PR TITLE
006 instrument interface

### PR DESCRIPTION
**Pull request by WojciechRynczuk**
The line counter has been integrated using the observer pattern. The observer uses the interface class called Instrument. So, the line counter is an implementation of the Instrument abstract class.

### CODE REVIEW

File: development/config/Doxyfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Doxyfile 1.8.12
+# Doxyfile 1.8.11
```
Comment (endrjux) [inline development/config/Doxyfile]: Why did you changed to lower version?

File: sources/common/inc/LineCounter.h [hunk 2]
```diff
@@ -95,13 +95,13 @@ namespace PARSER
             static const std::string HIGH_COUNTER_NAME;
 
             /// The counter name.
-            const std::string m_CounterName{};
+            const std::string m_CounterName {};
```
Comment (endrjux) [inline sources/common/inc/LineCounter.h]: Please revert all that changes...

File: sources/common/inc/LineCounter.h [hunk 3]
```diff
@@ -32,41 +32,38 @@
...
+                        SIGNAL::SourceRegistry &sourceRegistry,
+                        SIGNAL::SignalDb &signalDb);
 
-            /// Updates the line counting information.
-            ///
-            /// @param timestamp The time the signal has been registered.
-            /// @param lineNumber The line number referencing to the signal's occurrence.
-            void Update(uint64_t timestamp, LineNumberT lineNumber);
+            /// @copydoc Instrument::Notify()
+            void Notify(LineNumberT lineNumber, const SIGNAL::Signal &signal);
```
Comment (endrjux) [inline sources/common/inc/LineCounter.h]: `virtual`

File: sources/common/inc/LineCounter.h [hunk 4]
```diff
@@ -32,41 +32,38 @@
...
 
-            /// Adds line counter signal to given database.
-            ///
-            /// @param signalDb Signal database.
-            /// @param sourceHandle Signal source handle.
-            void RecordToSignalDb(SIGNAL::SignalDb &signalDb,
-                                  SIGNAL::SourceRegistry::HandleT sourceHandle);
+            /// @copydoc Instrument::Terminate()
+            void Terminate() const;
```
Comment (endrjux) [inline sources/common/inc/LineCounter.h]: `virtual`

File: sources/common/inc/Instrument.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            Instrument(SIGNAL::SourceRegistry &sourceRegistry,
+                       SIGNAL::SignalDb &signalDb,
+                       const std::string &sourceName);
+
+            /// The signal database.
+            SIGNAL::SignalDb &m_rSignalDb;
+
+            /// The instrument handle.
+            SIGNAL::SourceRegistry::HandleT m_InstrumentHandle;
+    };
```
Comment (endrjux) [inline sources/common/inc/Instrument.h]: Pure virtual destructor missing.

File: sources/common/inc/LogParser.h [hunk 6]
```diff
@@ -49,9 +51,15 @@ namespace PARSER
     {
         public:
 
-            /// Returns reference to the signal database.
+            /// Returns the const reference to the signal database.
             const SIGNAL::SignalDb &GetSignalDb() const
             {
+                return const_cast<SIGNAL::SignalDb &>((this)->GetSignalDb());
```
Comment (endrjux) [inline sources/common/inc/LogParser.h]: What is this abomination?
What is it needed for?
How is it now needed for something if it wasn't before?

File: sources/common/inc/LogParser.h [hunk 7]
```diff
@@ -61,6 +69,17 @@ namespace PARSER
                 return std::move(m_pSignalDb);
             }
 
+            /// Invokes the parser.
+            void Execute();
+
+            /// Attaches an instrument to the parser.
+            ///
+            /// @param instrument An instrument to be attached.
+            void Attach(INSTRUMENT::Instrument &instrument);
```
Comment (endrjux) [inline sources/common/inc/LogParser.h]: `rInstrument` I guess...

File: sources/common/inc/TxtParser.h [hunk 8]
```diff
@@ -64,9 +60,6 @@ namespace PARSER
             /// Parses the input file.
             void Parse();
```
Comment (endrjux) [inline sources/common/inc/TxtParser.h]: `virtual`. Also this is `protected` not `private`.

File: sources/common/inc/LogParser.h [hunk 9]
```diff
@@ -75,9 +94,15 @@ namespace PARSER
                       SIGNAL::SourceRegistry &sourceRegistry,
                       bool verboseMode);
 
+            /// Triggers parsing the source.
+            virtual void Parse() = 0;
```
Comment (endrjux) [inline sources/common/inc/LogParser.h]: Pure virtual destructor missing.

File: sources/common/src/Instrument.cpp [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+/// IN THE SOFTWARE.
+
+#include "Instrument.h"
+
+INSTRUMENT::Instrument::Instrument(SIGNAL::SourceRegistry &sourceRegistry,
+                                   SIGNAL::SignalDb &signalDb,
+                                   const std::string &sourceName) :
+    m_rSignalDb(signalDb)
+{
+    m_InstrumentHandle = sourceRegistry.Register(sourceName);
```
Comment (endrjux) [inline sources/common/src/Instrument.cpp]: This could be inlined in hpp.

File: sources/vcdMaker/src/vcdMaker.cpp [hunk 11]
```diff
@@ -45,15 +46,36 @@ int main(int argc, const char *argv[])
...
     {
-        // Parse the log file.
+        // Create the log parser.
         PARSER::TxtParser txtLog(cli.GetInputFileName(),
                                  cli.GetTimebase(),
                                  registry,
-                                 cli.GetLineCounterName(),
                                  cli.IsVerboseMode());
 
+        if (cli.GetLineCounterName().size())
```
Comment (endrjux) [inline sources/vcdMaker/src/vcdMaker.cpp]: Super bad. `std::string::empty()`.

File: sources/vcdMaker/src/vcdMaker.cpp [hunk 12]
```diff
@@ -45,15 +46,36 @@ int main(int argc, const char *argv[])
...
+        {
+            // Register the line counting instrument.
+            lineCounter = new INSTRUMENT::LineCounter(cli.GetInputFileName(),
+                    cli.GetLineCounterName(),
```
Comment (endrjux) [inline sources/vcdMaker/src/vcdMaker.cpp]: Indentation wrong.

File: sources/vcdMaker/src/vcdMaker.cpp [hunk 13]
```diff
@@ -45,15 +46,36 @@ int main(int argc, const char *argv[])
     // Source registry.
     SIGNAL::SourceRegistry registry;
 
+    // Line counter.
+    INSTRUMENT::LineCounter *lineCounter = NULL;
```
Comment (endrjux) [inline sources/vcdMaker/src/vcdMaker.cpp]: Use `std::unique_ptr` and `std::make_unique` if you must use pointer. I don't like this.

File: sources/vcdMaker/src/vcdMaker.cpp [hunk 14]
```diff
@@ -45,15 +46,36 @@ int main(int argc, const char *argv[])
...
         PARSER::TxtParser txtLog(cli.GetInputFileName(),
                                  cli.GetTimebase(),
                                  registry,
-                                 cli.GetLineCounterName(),
                                  cli.IsVerboseMode());
 
+        if (cli.GetLineCounterName().size())
+        {
+            // Register the line counting instrument.
+            lineCounter = new INSTRUMENT::LineCounter(cli.GetInputFileName(),
```
Comment (endrjux) [inline sources/vcdMaker/src/vcdMaker.cpp]: I don't like the idea. Create dynamically, pass as reference, then again store as poiner. Either use `std::unique_ptr` or `std::ref` or maybe even both...

File: sources/vcdMaker/src/vcdMaker.cpp [hunk 15]
```diff
@@ -45,15 +46,36 @@ int main(int argc, const char *argv[])
...
+        }
+
+        // Start parsing.
+        txtLog.Execute();
+
+        // Line counter is no longer needed.
+        if (lineCounter)
+        {
+            delete lineCounter;
+        }
```
Comment (endrjux) [inline sources/vcdMaker/src/vcdMaker.cpp]: This shouldn't be even needed...

File: sources/vcdMerge/src/SignalSource.cpp [hunk 16]
```diff
@@ -70,9 +71,30 @@ void MERGE::SignalSource::Create()
     PARSER::TxtParser parser(m_Filename,
                              m_TimeUnit,
                              m_rSignalRegistry,
-                             m_LineCounter,
                              m_VerboseMode);
 
+    // Line counter.
+    INSTRUMENT::LineCounter *lineCounter = NULL;
```
Comment (endrjux) [inline sources/vcdMerge/src/SignalSource.cpp]: Same comments as in previous file...

File: sources/common/src/LogParser.cpp [hunk 17]
```diff
@@ -44,3 +44,26 @@ PARSER::LogParser::LogParser(const std::string &filename,
                                  "' failed, it either doesn't exist or is inaccessible.");
     }
 }
+
+void PARSER::LogParser::Attach(INSTRUMENT::Instrument &instrument)
+{
+    m_vpInstruments.push_back(&instrument);
```
Comment (endrjux) [inline sources/common/src/LogParser.cpp]: I don't like this design, deatils in `vcdMaker.cpp` comments.

Comment (endrjux) [inline sources/common/src/LogParser.cpp]: Btw, it could be inline in hpp.

File: sources/common/src/LogParser.cpp [hunk 18]
```diff
@@ -44,3 +44,26 @@ PARSER::LogParser::LogParser(const std::string &filename,
...
+}
+
+void PARSER::LogParser::Execute()
+{
+    // Process the log.
+    Parse();
+
+    // Trigger the final action of instruments.
+    TerminateInstruments();
+}
```
Comment (endrjux) [inline sources/common/src/LogParser.cpp]: Comments above are useless. If you feel need for a comment in such place, change function name to better one.
Could be inlined in hpp.

File: development/config/Doxyfile (hunk 1 above)
Comment (endrjux) [inline development/config/Doxyfile]: I actually think that there is no need to modify this file in this PR, it's about something different.

File: sources/common/inc/LogParser.h (hunk 7 above)
Comment (WojciechRynczuk) [inline sources/common/inc/LogParser.h]: We weren't doing so but I like the idea. I will update this at least for the files I will be touching in this  issue.

File: sources/common/inc/Instrument.h [hunk 19]
```diff
@@ -68,9 +68,16 @@ namespace INSTRUMENT
...
             /// @param sourceName A source name.
-            Instrument(SIGNAL::SourceRegistry &sourceRegistry,
-                       SIGNAL::SignalDb &signalDb,
-                       const std::string &sourceName);
+            Instrument(SIGNAL::SourceRegistry &rSourceRegistry,
+                       SIGNAL::SignalDb &rSignalDb,
+                       const std::string &rSourceName) :
+                m_rSignalDb(rSignalDb)
+            {
+                m_InstrumentHandle = rSourceRegistry.Register(rSourceName);
```
Comment (endrjux) [inline sources/common/inc/Instrument.h]: This can just be done in init list like `m_rSignalDb`.

File: sources/common/src/TxtParser.cpp [hunk 20]
```diff
@@ -32,16 +32,16 @@
...
                              bool verboseMode) :
-    LogParser(filename, sourceRegistry, verboseMode),
+    LogParser(rFilename, rSourceRegistry, verboseMode),
     m_ValidLines(0),
     m_InvalidLines(0),
-    m_SourceHandle(sourceRegistry.Register(filename))
+    m_SourceHandle(rSourceRegistry.Register(rFilename))
 {
-    m_pSignalDb = std::make_unique<SIGNAL::SignalDb>(timeBase);
+    m_pSignalDb = std::make_unique<SIGNAL::SignalDb>(rTimeBase);
```
Comment (endrjux) [inline sources/common/src/TxtParser.cpp]: Hm, this can be in init list too I think.

File: sources/common/src/TxtParser.cpp [hunk 21]
```diff
@@ -32,38 +32,16 @@
...
-    // Process the log
-    Parse();
-
-    // Dump the line counting information.
-    if (m_LineCounterEnabled)
-    {
-        m_LineCounter.RecordToSignalDb(*(m_pSignalDb.get()),
-                                       m_LineCounterSourceHandle);
-    }
+    m_pSignalDb = std::make_unique<SIGNAL::SignalDb>(rTimeBase);
```
Comment (endrjux) [inline sources/common/src/TxtParser.cpp]: This can be in init list too...

File: sources/common/inc/LogParser.h (hunk 6 above)
Comment (MichalakMarcin) [inline sources/common/inc/LogParser.h]: What about this Andrzej's request? Do we leave it as it is now?

Comment (MichalakMarcin) [inline sources/common/inc/LogParser.h]: Actually it looks like you don't need currently the "const" version. Do you?

Comment (WojciechRynczuk) [inline sources/common/inc/LogParser.h]: Well, we agreed that two get methods are OK so as not to violate existing 'contracts'. E.g. VCDTrace constructor takes the constant reference as it doesn't change anything and only produces the output. On the other hand, the LineCounter needs the non-constant reference as it adds signals to the database.

File: sources/common/src/TxtParser.cpp (hunk 21 above)
Comment (WojciechRynczuk) [inline sources/common/src/TxtParser.cpp]: It couldn't be done there as the base class members cannot be directly initialized on the init list of the inheriting class. But I liked the idea and moved this initialization to the LogParser constructor's init list.

File: sources/vcdMaker/src/vcdMaker.cpp (hunk 14 above)
Comment (WojciechRynczuk) [inline sources/vcdMaker/src/vcdMaker.cpp]: It has been decided to leave it as it is. Containers store just objects. They cannot store references. So, in our case they store pointers. The reference is passed so as not to check against the NULL pointer. It should always be not NULL. The solution with std::ref worked but the syntax was confusing.

**Pull request merged**